### ISSUE TITLE
kernel_shape missing when exporting to onnx with dynamo=True

### ISSUE THREAD

**Issue by silverberry-rpi** [pytorch/pytorch#190890]
### 🐛 Describe the bug

This seems to be the same problem as reported in:
https://github.com/pytorch/pytorch/issues/169824
I'm using a onnx to hailo design flow and the missing kernel shape attribute breaks the hailo tools. I verified in netron the attribute kernel_shape is missing when exporting with dynamo=true and when dynamo=false its correctly present and the hailo tools work.
Here's an extra hint, the missing kernel shapes are associated with conv2d layers. When I created these layers I used the size parameter from the yolo.cfg file and passed that int (instead of a h,w tuple) to the layer create. This is interpreted by the conv2d class as a kernel of (size, size). But maybe this causes some confusion in the translator? Just a guess I haven't had time to investigate that yet.
Torch 2.13.0
Onnxscript 0.7.1
Onnx 1.22.0
Here's my call to export that suceeded in outputting the kernel_shape:
torch.onnx.export(model.float(),
                       (model_inputs,),
                       "yolov3.onnx",
                       input_names=["input"],
                       output_names=["output"],
                       export_params=True,
                       training=torch.onnx.TrainingMode.PRESERVE,
                       do_constant_folding=False,
                       opset_version=17,
                       dynamo=False)
The call is full of deprecated arguments because in an effort to solve this problem I  copied it from hailo documentation which seems out of date. 

### Versions

Collectting environment information...
PyTorch version: 2.13.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-136-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A
CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  8
On-line CPU(s) list:                     0-7
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz
CPU family:                              6
Model:                                   85
Thread(s) per core:                      1
Core(s) per socket:                      1
Socket(s):                               8
Stepping:                                7
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep
_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpu
id_fault ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1
 xsaves arat umip pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               256 KiB (8 instances)
L1i cache:                               256 KiB (8 instances)
L2 cache:                                32 MiB (8 instances)

Versions of relevant libraries:
[pip3] numpy==2.5.1
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.3.33
[pip3] nvidia-nvtx==13.0.85
[pip3] onnx==1.22.0
[pip3] onnx-ir==0.2.1
[pip3] onnxruntime==1.27.0
[pip3] onnxscript==0.7.1
[pip3] onnxsim==0.6.5
[pip3] onnxslim==0.1.94
[pip3] optree==0.19.1
[pip3] torch==2.13.0
[pip3] torchvision==0.28.0
[pip3] triton==3.7.1
[conda] Could not collect

cc @justinchuby @titaiwangms @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

**Comment by justinchuby** [comment]
@titaiwangms

**Comment by titaiwangms** [comment]
@silverberry-rpi Can you export it with opset18? 17 is pretty old and was designed more for torchscript exporter. Also, it would be great if you can provide a full repro script (not model). I am pretty sure https://github.com/pytorch/pytorch/issues/169824 is not reproducible in nightly.

**Comment by silverberry-rpi** [comment]
Ok I'll work on providing a better report.  I reported using opset 17 because its at the upper range of what hailo reports supporting and tnis example reflects what I pulled from their documentation which I think is outdated. Previously I was using the more modern api for export with dynamo=true and specifying opset=21 and the kernel_shape attribute was not exported. I think the problem is not constrained by opset.  But I'll develop a minimal test case.  This will take me a couple days, I have a lot on my plate right now. 
To clarify do you want the test case to use the modern api and show the failure or use the deprecated api and show success? Please prioritize.

**Comment by silverberry-rpi** [comment]
Ok, I  got ai to help me ;-)
The single conv2d node in model.onnx has no kernel_shape attribute, in model2.onnx it is correctly present.  Let me know if you need anything else.

import torch
import torch.nn as nn

class MyModel(nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.conv = nn.Conv2d(1, 3, kernel_size=3, stride=1, padding=1)
    def forward(self, x):
        return self.conv(x)

model = MyModel()
model.eval()
x = torch.randn(1, 1, 224, 224)
torch.onnx.export(model, x, "model.onnx", export_params=True)
torch.onnx.export(model, x, "model2.onnx", export_params=True, dynamo=False)

**Comment by titaiwangms** [comment]
@silverberry-rpi Thanks! I misunderstood the issue. It's indeed a missing attribute. Out of curiosity, why is this a necessity for ONNX model usage as from spec I see `kernel_shape` can be inferred when it's not provided - https://github.com/onnx/onnx/blob/main/docs/Operators.md#Conv

**Comment by silverberry-rpi** [comment]
> [@silverberry-rpi](https://github.com/silverberry-rpi) Thanks! I misunderstood the issue. It's indeed a missing attribute. Out of curiosity, why is this a necessity for ONNX model usage as from spec I see `kernel_shape` can be inferred when it's not provided - https://github.com/onnx/onnx/blob/main/docs/Operators.md#Conv

This is useful. I can pass this on to the hailo support as a recomendation for improving the robustness of their parser/translator. As it stands right now if the kernel_shape attribute is missing, their python code crashes with a list invalid index exception. I patched my local venv copy to intercept the exception, but I  could also add some code to extrapolate the kernel shape from W.  That would patch up the design flow until a pytorch export fix is available. Thanks for this.

Hailo supports a parser/translator/quantization/synthesis flow that maps an onnx neural network onto hardware blocks in their npu accelerator asic.  This requires some rather strict rules so they expect the model to meet certain constraints. They also support tensorflow but not pytorch directly. They require a translation from pytorch to onnx. A direct pytorch translation would be a welcome improvement. 

You can read more about this at:
https://hailo.ai/developer-zone/documentation/dataflow-compiler-v3-34-0/?sp_referrer=sdk/translating_tf_models.html
This might require registration to access. Its worth it for the concentrated information. 

There is a short reference to pytorch. Note their use of deprecated interfaces. I should report this to them, once there is a fix for this pytorch issue. The new dynamo exporter produces a much cleaner and simpler onnx translation.

```python
# Building a simple PyTorch model
class SmallExample(torch.nn.Module):
   def __init__(self):
      super(SmallExample, self).__init__()
      self.conv1 = torch.nn.Conv2d(96, 24, kernel_size=(1, 1), stride=(1, 1))
      self.bn1 = torch.nn.BatchNorm2d(24)
      self.relu1 = torch.nn.ReLU6()

   def forward(self, x):
      x = self.conv1(x)
      x = self.bn1(x)
      x = self.relu1(x)
      return x

# Exporting the model to ONNX
torch_model = SmallExample()
torch_model.eval()
inp = [torch.randn((1, 96, 24, 24), requires_grad=False)]
torch_model(*inp)
onnx_path = 'small_example.onnx'

# Note the used args:
# export_params makes sure the weight variables are part of the exported ONNX,
# training=TrainingMode.PRESERVE preserves layers and variables that get folded into other layers in EVAL mode (inference),
# do_constant_folding is a recommendation by pytorch to prevent issues with PRESERVED mode,
# opset_version selects the desired ONNX implementation (currently Hailo support opset versions 8 and 11-17).
torch.onnx.export(torch_model, tuple(inp), onnx_path,
                  export_params=True,
                  training=torch.onnx.TrainingMode.PRESERVE,
                  do_constant_folding=False,
                  opset_version=13)
```

Thanks for your support.

**Comment by titaiwangms** [comment]
Nightly onnxscript should fix this - https://github.com/microsoft/onnxscript/pull/2972

**Comment by silverberry-rpi** [comment]
> Nightly onnxscript should fix this - [microsoft/onnxscript#2972](https://github.com/microsoft/onnxscript/pull/2972)

Thanks for providing this. I'd like to test it.  I ran:
pip install --upgrade --pre --index-url https://aiinfra.pkgs.visualstudio.com/PublicPackages/_packaging/ORT-Nightly/pypi/simple/ onnxscript
But seem to have picked up this:
pip list | grep onnxscript
onnxscript                   0.7.2.dev20260714
Obviously the fix is not in that version.  Can someone give me the correct pip command? Thanks.

**Comment by justinchuby** [comment]
Please try https://pypi.org/project/onnxscript/0.7.2.dev20260725/

**Comment by silverberry-rpi** [comment]
> Please try https://pypi.org/project/onnxscript/0.7.2.dev20260725/

Thanks, that worked.
Exported my pytorch model to onnx with dynamo=true and the kernel_shape attribute was present on the conv2d layers as verified in netron. The exported model passed through the hailo tools without error.  Problem solved.
Thank you.
